dockerExtension.ts

```diff
@@ -281,6 +297,90 @@ namespace Configuration {
+                new DefaultAppStorageProvider(fileSystemProvider),
+                new DefaultDebuggerClient(
+                    new RemoteVsDbgClient(
+                        dockerOutputManager,
+                        fileSystemProvider,
+                        ctx.globalState,
+                        osProvider,
+                        processProvider)),
+                dockerClient,
+                dockerOutputManager,
+                fileSystemProvider,
+                osProvider,
+                processProvider,
+                ctx.workspaceState);
+
+        ctx.subscriptions.push(
+            vscode.debug.registerDebugConfigurationProvider(
+                'docker-netcoreapp',
+                new DockerDebugConfigurationProvider(
+                    dockerManager,
+                    fileSystemProvider,
+                    osProvider,
+                    new MsBuildNetCoreProjectProvider(
+                        fileSystemProvider,
+                        new CommandLineMSBuildClient(processProvider),
+                        new OSTempFileProvider(
+                            osProvider,
+                            processProvider)),
+                    async () => {
+                        // NOTE: Debugging .NET Core in Docker containers requires the C# (i.e. .NET Core debugging) extension.
+                        //       As Docker debugging is experimental, we don't want the extension as a whole to depend on it.
+                        //       Hence, we only check for its existence if/when asked to debug .NET Core in Docker containers.
+                        // TODO: Remove this check once debugging is enabled-by-default (and then add a package.json extension dependency).
+                        const dependenciesSatisfied = vscode.extensions.getExtension('ms-vscode.csharp') !== undefined;
+
+                        if (!dependenciesSatisfied) {
+                            const openExtensionInGallery: vscode.MessageItem = {
+                                title: 'View extension in gallery'
+                            };
+
+                            const result = await vscode.window.showErrorMessage(
+                                'To debug .NET Core in Docker containers, install the C# extension for VS Code.',
+                                openExtensionInGallery);
+
+                            if (result === openExtensionInGallery) {
+                                const browserClient: BrowserClient = new OpnBrowserClient();
+
+                                browserClient.openBrowser('https://marketplace.visualstudio.com/items?itemName=ms-vscode.csharp');
+                            }
+                        }
+
+                        return dependenciesSatisfied;
+                    })));
+
+        ctx.subscriptions.push(
+            vscode.debug.onDidChangeActiveDebugSession(
+                session => {
+                    if (session === undefined) {

```

Comment:
The intent is to shutdown any containers used for Docker debugging still running when debugging ends in VS Code, whether from the ending session or a previous one.  I'm not sure it matters which kind of session was ending, in that if there are no containers running it's a no-op and if there are, then we'd still want them shutdown.

(As far as I know, VS Code only supports one debugging session at a time.)

Reply:
I'm just concerned because this code will be run for _every_ user that has the docker extension installed - even if they never touch the 'docker-netcoreapp' experience (well - every user once we get rid of the feature flag). No matter how perfect the code is today (which could change in the future) I would prefer to have some checking here to prevent the docker extension from affecting users that are debugging entirely different things.

Can you use `onDidTerminateDebugSession` and check `session.type === 'docker-netcoreapp'`?

Reply:
I ran into issues with `onDidTerminateDebugSession` when the VS Live Share extension was installed, as it may "wrap" other debug adaptors, even when not actively used.  In that case, the returned `type` is the Live Share adaptor.  It exposes the underlying adaptor type via a custom request, but when a session is terminated you cannot make custom requests.

Instead, I add a "session manager" that's informed by my debug adaptor when debugging is about to start, at which time it starts listening to the termination event.  Then, on termination, it performs the necessary cleanup and stops listening for future events.  This should limit it the cleanup to running only when the user has attempted to debug within a Docker container, but no more than once per attempt.